skipruntime-ts/helpers/src/utils.ts

```diff
@@ -68,3 +71,96 @@ export function reactiveResponseHeader(
 ): [string, string] {
   return ["Skip-Reactive-Response-Token", JSON.stringify(reactiveResponse)];
 }
+
+/*****************************************************************************/
+// Joins
+/*****************************************************************************/
+
+type JoinObject = { value: JsonObject; side: "left" | "right" };
+
+class AddJoinSideField implements Mapper<Json, Json, Json, JoinObject> {
+  constructor(private joinSide: "left" | "right") {}
+
+  mapEntry(
+    key: Json,
+    values: NonEmptyIterator<JsonObject>,
+  ): Iterable<[Json, JoinObject]> {
+    const result: [Json, JoinObject][] = [];
+    for (const v of values) {
+      result.push([key, { value: v, side: this.joinSide }]);
+    }
+    return result;
+  }
+}
+
+function mergeObjects(object1: JoinObject, object2: JoinObject): Json {
+  const v1 = object1.value;
+  const v2 = object2.value;
+  const keys1 = Object.keys(v1);
+  if (keys1.some((key) => key in v2)) {
+    throw new Error(
+      "Objects don't have distinct fields: " +
+        JSON.stringify(v1) +
+        " " +
+        JSON.stringify(v2),
+    );
```

Comment:
A question on the error behavior: join in relational algebra drops pairs that do not agree on common fields, while here it is an error for there to be common fields (other than the key). No strong position, but it isn't immediately obvious to me that this choice is the one users will expect, or be the most useful.

Reply:
To clarify, I wonder about just dropping the pairs with common fields other than the key, rather than throwing, not taking pairs that happen to have matching values for common fields.

Reply:
I don't know people know relational algebra well enough to expect that. I suspect they don't.
I am happy to add an option (dropOnCommonFields for example), but making that the default is not a good idea IMO. Imagine that is not the behavior you expect, now imagine you end up inserting new kinds of objects in a collection with a field that collides with another collection by mistake. The debugging is going to be pretty brutal ...

Reply:
Ok makes sense, the debugging angle is compelling. And we can wait to see if anyone asks to drop instead of error.